pkg/cpio/fs_unix.go

```diff
@@ -0,0 +1,152 @@
+	modeSGID        = 0002000
+	modeSticky      = 0001000
+	modePermissions = 0000777
+)
+
+var modeMap = map[uint64]os.FileMode{
+	modeSocket:  os.ModeSocket,
+	modeSymlink: os.ModeSymlink,
+	modeFile:    0,
+	modeBlock:   os.ModeDevice,
+	modeDir:     os.ModeDir,
+	modeChar:    os.ModeCharDevice,
+	modeFIFO:    os.ModeNamedPipe,
+}
+
+// modes sets the modes, changing the easy ones first and the harder ones last.
+// In this way, we set as much as we can before bailing out. It's not an error
+// to not be able to set uid and gid, at least not yet.
+// For now we also ignore sticky bits.
+func modes(r Record) error {
+	if err := os.Chmod(r.Name, os.FileMode(r.Mode)); err != nil {
+		return err
+	}
+	if err := os.Chtimes(r.Name, time.Now(), time.Unix(int64(r.MTime), 0)); err != nil {
+		return err
+	}
+	if err := os.Chown(r.Name, int(r.UID), int(r.GID)); err != nil {
+		return err
+	}
+	// TODO: only set SUID and GUID if we can set the owner.
+	return nil
+}
+
+func perm(r Record) uint32 {
+	return uint32(r.Mode) & modePermissions
+}
+
+func dev(r Record) int {
+	return int(r.Rmajor<<8 | r.Rminor)
+}
+
+func linuxModeToMode(m uint64) (os.FileMode, error) {
+	if t, ok := modeMap[m&modeTypeMask]; ok {
+		return t, nil
+	}
+	return 0, fmt.Errorf("Invalid file type %#o", m&modeTypeMask)
+}
+
+func CreateFile(f Record) error {
+	m, err := linuxModeToMode(f.Mode)
+	if err != nil {
+		return err
+	}
+
+	switch m {
+	case os.ModeSocket, os.ModeNamedPipe:
+		return fmt.Errorf("%q: type %v: we do not create IPC endpoitns", f.Name, m)
+	case os.FileMode(0):
+		nf, err := os.Create(f.Name)
+		if err != nil {
+			return err
+		}
+		defer nf.Close()
+		if _, err := io.Copy(nf, f); err != nil {
+			return err
+		}
+		return modes(f)
+	case os.ModeDir:
+		if err := os.MkdirAll(f.Name, os.FileMode(f.Mode)); err != nil {
+			return err
+		}
+		return modes(f)
+	case os.ModeDevice:
+		if err := syscall.Mknod(f.Name, perm(f)|syscall.S_IFBLK, dev(f)); err != nil {
+			return err
+		}
+		return modes(f)
+	case os.ModeCharDevice:
+		if err := syscall.Mknod(f.Name, perm(f)|syscall.S_IFCHR, dev(f)); err != nil {
+			return err
+		}
+		return modes(f)
+	case os.ModeSymlink:
+		// TODO
+		return nil
+	default:
+		return fmt.Errorf("%v: Unknown type %#o", f.Name, m)
+	}
+}
+
+func GetRecord(path string) (Record, error) {
+	fi, err := os.Lstat(path)
+	if err != nil {
+		return Record{}, err
+	}
+	sys := fi.Sys().(*syscall.Stat_t)
+	info := Info{
+		Ino:      sys.Ino,
+		Mode:     uint64(sys.Mode),
+		UID:      uint64(sys.Uid),
+		GID:      uint64(sys.Gid),
+		NLink:    sys.Nlink,
+		MTime:    uint64(sys.Mtim.Sec),
+		FileSize: uint64(sys.Size),
+		Major:    sys.Dev >> 8,
+		Minor:    sys.Dev & 0xff,
+		Rmajor:   sys.Rdev >> 8,
+		Rminor:   sys.Rdev & 0xff,
+		Name:     path,
+	}
+	switch {
+	case fi.Mode().IsRegular():
+		osfile, err := os.Open(path)
+		if err != nil {
+			return Record{}, err
+		}
+		return Record{osfile, info}, nil
+	case fi.Mode()&os.ModeSymlink != 0:
```

Comment:
it's a lot safer (since they have some 2 bit fields in their mode descriptions already where this test would lead to a false positive -- see ModeChardevice) to do fi.Mode()&os.ModeSymlink == os.ModesSymlink. Better safe than sorry later. Honestly, the best way is to & with ModeType and compare to ModeSymlink.